lib/srv/desktop/windows_server.go

```diff
@@ -660,7 +752,45 @@ func (s *WindowsService) connectRDP(ctx context.Context, log logrus.FieldLogger,
 	ctx, cancel := context.WithCancel(ctx)
 	defer cancel()
 
+	delay := timer()
 	tdpConn := tdp.NewConn(proxyConn)
+	sessionStartTime := s.cfg.Clock.Now().UTC().Round(time.Millisecond)
+	tdpConn.OnSend = func(m tdp.Message, b []byte) {
+		switch b[0] {
+		case byte(tdp.TypePNGFrame), byte(tdp.TypeClipboardData):
+			if err := sw.EmitAuditEvent(ctx, &events.DesktopRecording{
+				Metadata: events.Metadata{
+					Type: libevents.DesktopRecordingEvent,
+					Time: s.cfg.Clock.Now().UTC().Round(time.Millisecond),
+				},
+				Message:           b,
+				DelayMilliseconds: delay(),
+			}); err != nil {
+				s.cfg.Log.WithError(err).Warning("could not emit desktop recording event")
+			}
+		}
+	}
+	tdpConn.OnRecv = func(m tdp.Message) {
+		switch m.(type) {
+		// TODO: tdp.ClipboardData
+		case tdp.ClientScreenSpec, tdp.MouseButton, tdp.MouseMove:
+			b, err := m.Encode()
+			if err != nil {
+				s.cfg.Log.WithError(err).Warning("could not emit desktop recording event")
```

Comment:
Should we record the fact that the encode operation failed? It is worth having that and/or the error in the event?

Reply:
As of now, this can basically never happen, as encode just writes to a `*bytes.Buffer`, which is guaranteed to always return a `nil` error.